addons/html_editor/static/src/core/history_plugin.js

```diff
@@ -404,77 +436,137 @@ export class HistoryPlugin extends Plugin {
+                filteredRecords.push(record);
                 continue;
             }
-            filteredRecords.push(record);
+            // Add entry for current target if not already present.
+            if (!nodeToAttributes.has(record.target)) {
+                nodeToAttributes.set(record.target, new Set());
+            }
+            const visitedAttributes = nodeToAttributes.get(record.target);
+            // Keep only the first mutation record for each attribute.
+            if (!visitedAttributes.has(record.attributeName)) {
+                filteredRecords.push(record);
+                visitedAttributes.add(record.attributeName);
+            }
         }
-        // @todo @phoenix allow an option to filter mutation records.
         return filteredRecords;
     }
 
+    /**
+     * Class attribute records are expanded into multiple classList records.
+     * Attribute records have their oldValue normalized and newValue added to it.
+     * @todo: expand childList mutations to add/remove records.
+     *
+     * @param { MutationRecord } record
+     * @returns { HistoryMutationRecord | HistoryMutationRecord[] }
+     */
+    transformRecord(record) {
+        if (record.type === "attributes") {
+            if (record.attributeName === "class") {
+                return this.splitClassMutationRecord(record);
+            }
+            const oldValue = record.oldValue === undefined ? null : record.oldValue;
+            const newValue = record.target.getAttribute(record.attributeName);
+            const { type, target, attributeName } = record;
+            return { type, target, attributeName, oldValue, newValue };
+        }
+        return record;
+    }
+
+    /**
+     * Breaks down a class attribute mutation into individual class
+     * addition/removal records for more precise history tracking.
+     *
+     * @param { MutationRecord } record of type "attributes" with attributeName === "class"
+     * @returns { MutationRecordClassList[]}
+     */
+    splitClassMutationRecord(record) {
+        // oldValue can be nullish, or have extra spaces
+        const oldValue = record.oldValue?.split(" ").filter(Boolean);
+        const classesBefore = new Set(oldValue);
+        const classesAfter = new Set(record.target.classList);
+        // @todo: use Set.prototype.difference when it becomes widely available
+        const setDifference = (setA, setB) => {
+            const diff = new Set(setA);
+            setB.forEach((item) => diff.delete(item));
+            return diff;
+        };
+        const addedClasses = setDifference(classesAfter, classesBefore);
+        const removedClasses = setDifference(classesBefore, classesAfter);
+
+        /** @type {(className: string, operation: string) => MutationRecordClassList } */
+        const createClassRecord = (className, isAdded) => ({
+            type: "classList",
+            target: record.target,
+            className,
+            value: isAdded,
```

Comment:
@Goaman I have changed the oldValue/newValue thing for now, otherwise it would look very silly if the code that's supposed to arrive later never does...